Sources/PackageModel/PackageIdentity.swift

```diff
@@ -0,0 +1,178 @@
+        }
+
+        if string.removeUserComponentIfPresent() || detectedScheme != .ssh {
+            string.replaceFirstOccurenceIfPresent(of: ":", with: "/")
+        }
+        assert(!string.contains(":"))
+
+        var components = string.split(omittingEmptySubsequences: true, whereSeparator: isSeparator)
+
+        var lastPathComponent = components.popLast() ?? ""
+        lastPathComponent.removeSuffixIfPresent(".git")
+        components.append(lastPathComponent)
+
+        self.description = components.joined(separator: "/")
+        self.computedName = String(lastPathComponent)
+    }
+}
+
+extension PackageIdentity: Equatable {
+    public static func == (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.description == rhs.description
+    }
+}
+
+extension PackageIdentity: Comparable {
+    public static func < (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.description < rhs.description
+    }
+}
+
+extension PackageIdentity: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(self.description)
+    }
+}
+
+extension PackageIdentity: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let description = try container.decode(String.self)
+        self.init(description)
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.description)
+    }
+}
+
+extension PackageIdentity: JSONMappable, JSONSerializable {
+    public init(json: JSON) throws {
+        guard case .string(let string) = json else {
+            throw JSON.MapError.typeMismatch(key: "", expected: String.self, json: json)
+        }
+
+        self.init(string)
+    }
+
+    public func toJSON() -> JSON {
+        return .string(self.description)
+    }
+}
+
+extension PackageIdentity: ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self.init(value)
+    }
+}
+
+// MARK: -
+
+private enum Scheme: String, CustomStringConvertible, CaseIterable {
+    case ssh
+    case https
+    case git
+
+    var defaultPort: Int {
+        switch self {
+        case .ssh:
+            return 22
+        case .https:
+            return 443
+        case .git:
+            return 9418
+        }
+    }
+
+    public var description: String {
+        return self.rawValue
+    }
+}
+
+#if os(Windows)
+fileprivate let isSeparator: (Character) -> Bool = { $0 == "/" || $0 == "\\" }
+#else
+fileprivate let isSeparator: (Character) -> Bool = { $0 == "/" }
+#endif
+
+private extension StringProtocol where Self == Self.SubSequence {
+    @discardableResult
+    mutating func removeSuffixIfPresent<T: StringProtocol>(_ suffix: T) -> Bool {
+        guard hasSuffix(suffix) else { return false }
+        removeLast(suffix.count)
+        return true
+    }
+}
+
+private extension String {
```

Comment:
seems generally useful and maybe we can move elsewhere?

Reply:
I'm hesitant to add public extensions on core types, but if you see an opportunity for reuse across the project, I think that'd be something to consider in a separate PR.